perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_eduPersonScopedAffiliations.java

```diff
@@ -107,12 +107,61 @@ public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDe
 	}
 
 	@Override
-	public List<Pattern> getPatternsForMatch() {
-		List<Pattern> patterns = super.getPatternsForMatch();
-		patterns.add(userAllAttrsRemovedPattern);
-		patterns.add(userEPSAMARemovePattern);
-		patterns.add(userEPSAMASetPattern);
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = super.resolveVirtualAttributeValueChange(perunSession, message);
+
+		if (message instanceof AllAttributesRemovedForUser) {
+			log.debug("Resolving virtual attribute value change for message: " + message);
+			resolvingMessages.addAll(resolveAllAttributesRemovedForUser(perunSession, (AllAttributesRemovedForUser) message));
+		} else if (message instanceof AttributeSetForUser) {
+			log.debug("Resolving virtual attribute value change for message: " + message);
+			resolvingMessages.addAll(resolveAttributeSetForUser(perunSession, (AttributeSetForUser) message));
+		} else if (message instanceof AttributeRemovedForUser) {
+			log.debug("Resolving virtual attribute value change for message: " + message);
+			resolvingMessages.addAll(resolveAttributeRemovedForUser(perunSession, (AttributeRemovedForUser) message));
+		}
+
+		if (!resolvingMessages.isEmpty()) {
+			log.debug("These new messages will be generated: " + resolvingMessages);
+		}
+		return resolvingMessages;
+	}
+
+	private List<AuditEvent> resolveAllAttributesRemovedForUser(PerunSessionImpl perunSession, AllAttributesRemovedForUser allAttributesRemovedForUser) throws WrongAttributeAssignmentException, InternalErrorException, AttributeNotExistsException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		User user = allAttributesRemovedForUser.getUser();
+		Attribute attribute = perunSession.getPerunBl().getAttributesManagerBl().getAttribute(perunSession, user, getDestinationAttributeFriendlyName());
```

Comment:
I think this is not implemented correctly. This module should check if one of the three events that occured has attribute with friendly name `eduPersonScopedAffiliations` (the event for all attributes removed should always be processed).
If so, then it should try to get the current value of this virtual attribute. If the value is null, then you should return event AttributeRemovedForUser. If the value is non-null, you should return AttributeSetForUser.

Also, `getDestinationAttributeFriendlyName` doesn't return the whole urn, only the friendly name. All of those calls would fail. You need to pass the whole `urn:perun:user:......:eduPersonScopedAffiliations.